### PR TITLE
fix for documentation building and various other small accumulated changes

### PR DESCRIPTION
**Pull request by MatthiasHeilManchester**
- hot fix for building the documentation (so this should be moved across to main a.s.a.p.
- tweaked some of the build machinery for third party libraries (not sure why that stuff ever worked, but it's on the way out anyway (phew)
- tidied up some of @PaulAlexander23 crimes in the eigensolver code. Paul: Please have a look!

### CODE REVIEW

**Review by puneetmatharu — CHANGES_REQUESTED** [review]
Can't see any major issues; just a small suggestion for a tidy-up of the documentation-building workflow.

As a side note, it would be useful *ahem* if you were to provide links to the corresponding self-test runs to make sure we can check them easily. I've had a quick look and it's rather intriguing that the Ubuntu build time is 6x greater than the macOS ones despite having to run with only 2x less jobs. Is this likely to be down to the build of CGAL with the library? (If memory serves, I think we don't use CGAL on macOS.) It would be nice to see if we could install it with a package manager and try to get that build down from an hour to a 20mins (thinking of the poor penguins and seals, of course). I'm not too fussed about the actual validation step, it seems like the time difference between the two platforms is reasonable (43m on macOS vs 1h 47m on Ubuntu).

Will wait for the clean-up and @PaulAlexander23 to review the changes to his own code before merging stuff in.

File: self_test/mpi/eigen_solver/eigen_solver_test.cc
Comment (PaulAlexander23) [inline self_test/mpi/eigen_solver/eigen_solver_test.cc:484]: Same here. Indenting.

File: self_test/mpi/eigen_solver/eigen_solver_test.cc
Comment (PaulAlexander23) [inline self_test/mpi/eigen_solver/eigen_solver_test.cc:506]: Same here. Indenting.

**Review by PaulAlexander23 — CHANGES_REQUESTED** [review]
The changes made have caused some interesting indentation patterns. I would have fixed them myself but don't really know if you can pull request a pull request.

**Comment by PaulAlexander23** [comment]
Maybe we could also add clearer doxygen instructions in the coding convention. I don't think the single line triple slash comments before each variable individually, is mentioned.

File: self_test/generic/eigen_solver_test/eigen_solver_test.cc
Comment (puneetmatharu) [reply self_test/generic/eigen_solver_test/eigen_solver_test.cc:304]: Formally, we only format files in the `src/` directory but I agree that this indentation looks a little messy. @MatthiasHeilManchester would you be able to run `clang-format` on the file manually? IIRC, the command is something along the lines of
```bash
# clang-format -c <config-file> <target-file>
clang-format -c <oomph-root>/.clang-format self_test/generic/eigen_solver_test/eigen_solver_test.cc
```

**Comment by puneetmatharu** [comment]
> Maybe we could also add clearer doxygen instructions in the coding convention. I don't think the single line triple slash comments before each variable individually, is mentioned.

Had a look through the [Creating documentation](https://oomph-lib.github.io/oomph-lib/doc/creating_doc/html/index.html) page and can't find it either; I think we've only ever been told about this issue through word-of-mouth. Oops... Well spotted! (And I agree that we should add it.) 

However, if you are creating a self-test or demo driver (well, anything that will have auto-generated documentation), I would urge you @PaulAlexander23 to build the documentation yourself to make sure that the documentation for your driver works correctly.

File: .github/workflows/build-and-publish-docs.yaml
Comment (MatthiasHeilManchester) [reply .github/workflows/build-and-publish-docs.yaml]: done (though one would only scroll back if one knew it was there!). A lot of still learn git by watching, so I thought this was useful but I agree that it's tidier with out.

File: self_test/generic/eigen_solver_test/eigen_solver_test.cc
Comment (MatthiasHeilManchester) [reply self_test/generic/eigen_solver_test/eigen_solver_test.cc:304]: done -- sort of. I've done the emacs formatting on the entire file (both eigensolver codes) from within emacs. At least it's now consistent.

File: self_test/generic/eigen_solver_test/eigen_solver_test.cc
Comment (MatthiasHeilManchester) [reply self_test/generic/eigen_solver_test/eigen_solver_test.cc:406]: done

File: self_test/generic/eigen_solver_test/eigen_solver_test.cc
Comment (MatthiasHeilManchester) [reply self_test/generic/eigen_solver_test/eigen_solver_test.cc:478]: done

File: self_test/generic/eigen_solver_test/eigen_solver_test.cc
Comment (MatthiasHeilManchester) [reply self_test/generic/eigen_solver_test/eigen_solver_test.cc:503]: done

File: self_test/generic/eigen_solver_test/eigen_solver_test.cc
Comment (MatthiasHeilManchester) [reply self_test/generic/eigen_solver_test/eigen_solver_test.cc:543]: done

File: self_test/mpi/eigen_solver/eigen_solver_test.cc
Comment (MatthiasHeilManchester) [reply self_test/mpi/eigen_solver/eigen_solver_test.cc:275]: done

File: self_test/mpi/eigen_solver/eigen_solver_test.cc
Comment (MatthiasHeilManchester) [reply self_test/mpi/eigen_solver/eigen_solver_test.cc:314]: done

File: self_test/mpi/eigen_solver/eigen_solver_test.cc
Comment (MatthiasHeilManchester) [reply self_test/mpi/eigen_solver/eigen_solver_test.cc:394]: done

File: self_test/mpi/eigen_solver/eigen_solver_test.cc
Comment (MatthiasHeilManchester) [reply self_test/mpi/eigen_solver/eigen_solver_test.cc:484]: done

**Comment by MatthiasHeilManchester** [comment]
> > Maybe we could also add clearer doxygen instructions in the coding convention. I don't think the single line triple slash comments before each variable individually, is mentioned.
> 
> Had a look through the [Creating documentation](https://oomph-lib.github.io/oomph-lib/doc/creating_doc/html/index.html) page and can't find it either; I think we've only ever been told about this issue through word-of-mouth. Oops... Well spotted! (And I agree that we should add it.)
> 
> However, if you are creating a self-test or demo driver (well, anything that will have auto-generated documentation), I would urge you @PaulAlexander23 to build the documentation yourself to make sure that the documentation for your driver works correctly.

Agreed. I've added some text to doc/coding_conventions (but haven't actually looked at the doxygened version since "make" doesn't work  on that branch and I don't know how to build the documentation (may not even be done yet).

I did find an instructruction saying:

"Indentation of blocks etc. should follow the emacs standards."

which is true.  Is there a more formal way of describing this? What does clang-format actually do?

**Comment by puneetmatharu** [comment]
> Agreed. I've added some text to doc/coding_conventions (but haven't actually looked at the doxygened version yet. Will do this next time I have a makefile based build.

Bril, thanks!

> I did find an instruction saying:
> 
> "Indentation of blocks etc. should follow the emacs standards."
> 
> which is true. Is there a more formal way of describing this? What does clang-format actually do?

Hmmmm, I don't use Emacs *ahem* so I wouldn't have a strict definition of what you meant there. `clang-format` will format accordingly to the style guidelines we set in `.clang-format`. I'm starting to think that we should update the GitHub Actions `clang-format` workflow so that we format C++ code whenever it is part of a PR. That way we'll avoid annoying reformatting changes in future PRs to the same file. Thoughts? If you agree, I can look into this

**Comment by MatthiasHeilManchester** [comment]
> > Agreed. I've added some text to doc/coding_conventions (but haven't actually looked at the doxygened version yet. Will do this next time I have a makefile based build.
> 
> Bril, thanks!
> 
> > I did find an instruction saying:
> > "Indentation of blocks etc. should follow the emacs standards."
> > which is true. Is there a more formal way of describing this? What does clang-format actually do?
> 
> Hmmmm, I don't use Emacs _ahem_ so I wouldn't have a strict definition of what you meant there. `clang-format` will format accordingly to the style guidelines we set in `.clang-format`. I'm starting to think that we should update the GitHub Actions `clang-format` workflow so that we format C++ code whenever it is part of a PR. That way we'll avoid annoying reformatting changes in future PRs to the same file. Thoughts? If you agree, I can look into this



The (good!) reason why we didn't automatically clang format the demo drivers was that layout changes there can break the doxygen documentation in hard-to-spot ways. There are a couple of scripts that check for patterns in the generated html that (likely) indicate broken stuff, but it's not totally reliable and the only way to check (following running clang format on all the codes) would be for somebody to click through all the tutorials to check. 

  "do what emacs does" comes from the (good?) old days when shiny new students were given Andrew's .emacs file and blocks were formatted by highlighting them and pressing the tab key. Long before bloody vscode was invented.  I still do it that way...

**Comment by puneetmatharu** [comment]
> The (good!) reason why we didn't automatically clang format the demo drivers was that layout changes there can break the doxygen documentation in hard-to-spot ways. There are a couple of scripts that check for patterns in the generated html that (likely) indicate broken stuff, but it's not totally reliable and the only way to check (following running clang format on all the codes) would be for somebody to click through all the tutorials to check.

Ah yes, I was naively just thinking about non-demo-driver files, so clang-formatting all modified files is also a no-go. Hmm...

Maybe this is a discussion for oomph-lunch?

**Comment by MatthiasHeilManchester** [comment]
OK, are we done? @PaulAlexander23 , I think Puneet wanted you to have a look at my changes to your code to explicitly confirm that you're happy with my changes (modulo indentation which I've now fixed). Would be good if we could get the pull request in, and, if it passes also upgrade the main branch (though we should possibly try to get the bottom of the mac failures first).

**Review by PaulAlexander23 — APPROVED** [review]
Sure. I don't like the new formatting and the large number of changes between the files it produces, but yep, at least it is consistent. I am happy with the changes to the comments. Thanks.

**Comment by puneetmatharu** [comment]
Formatting is likely to change and will likely change if someone runs `clang-format` on Paul's source code but in the interest of getting this stuff in as soon as possible, and since the formatting doesn't "break" anything, I'm going to merge this in now.

**Comment by puneetmatharu** [comment]
@MatthiasHeilManchester please feel free to create a PR to merge the `development` branch into the `main` branch

**Pull request merged**
